driver-core/src/main/com/mongodb/operation/AsyncQueryBatchCursor.java

```diff
@@ -54,7 +55,7 @@
 import static java.lang.String.format;
 import static java.util.Collections.singletonList;
 
-class AsyncQueryBatchCursor<T> implements AsyncAggregateResponseBatchCursor<T> {
+class AsyncQueryBatchCursor<T> extends AbstractReferenceCounted implements AsyncAggregateResponseBatchCursor<T> {
```

Comment:
We shouldn't add a new public base class to achieve what should really be an internal invariant of this class. We don't want or need uses of this class to retain/release references.  Also, the class is already `Closeable`, and that really conflicts with the whole idea of reference counting.  There should only ever be one in progress operation on an instance, so I think we can achieve the goal by adding a single new boolean field tracking whether an operation is currently in progress.  Then we need to add the appropriate synchronization between in progress operation and the close method to ensure the invariants we are trying to enforce, which are:

* The resources (the server's cursor, and the connection source) are _eventually_ cleaned up after close() is called or the cursor is exhausted
* There is only one in progress operation 

In practice, that means that when close() is called, it should set the close bit. If there is no in progress operation, it should release its resources. Otherwise, it should do nothing else.  And when an operation completes, it should also release the instances's resource if the close bit is set (essentially completing the work of closing the cursor).

Reply:
Since you point out that this class and `AsyncChangeStreamBatchCursor` are `Closeable`, I have removed extending `AbstractReferenceCounted` and use an `AtomicBoolean` field to track operations in progress in both classes. When `close()` is called, `isClosed` is set to true, but resources are only released if there is no operation in progress.